### PR TITLE
Prepare for stabilizing libcore

### PR DESCRIPTION
**Pull request by alexcrichton**
These commits move libcore into a state so that it's ready for stabilization, performing some minor cleanup:
- The primitive modules for integers in the standard library were all removed from the source tree as they were just straight reexports of the libcore variants.
- The `core::atomic` module now lives in `core::sync::atomic`. The `core::sync` module is otherwise empty, but ripe for expansion!
- The `core::prelude::v1` module was stabilized after auditing that it is a subset of the standard library's prelude plus some primitive extension traits (char, str, and slice)
- Some unstable-hacks for float parsing errors were shifted around to not use the same unstable hacks (e.g. the `flt2dec` module is now used for "privacy").

After this commit, the remaining large unstable functionality specific to libcore is:
- `raw`, `intrinsics`, `nonzero`, `array`, `panicking`, `simd` -- these modules are all unstable or not reexported in the standard library, so they're just remaining in the same status quo as before
- `num::Float` - this extension trait for floats needs to be audited for functionality (much of that is happening in #27823)  and may also want to be renamed to `FloatExt` or `F32Ext`/`F64Ext`.
- Should the extension traits for primitives be stabilized in libcore?

I believe other unstable pieces are not isolated to just libcore but also affect the standard library.

cc #27701

### CODE REVIEW

File: src/libcoretest/num/flt2dec/mod.rs [hunk 1]
```diff
@@ -8,6 +8,10 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+// this is in principle private which is only made public for testing. do not
+// expose us in the documentation as a result
+#![doc(hidden)]
```
Comment (eefriedman) [inline src/libcoretest/num/flt2dec/mod.rs]: What's the point of this?  Nobody is ever going to generate docs for libcoretest.

Comment (alexcrichton) [inline src/libcoretest/num/flt2dec/mod.rs]: Oh oops I thought this was libcore, looks like this already exists in libcore!

**Pull request merged**